internal/pkg/service/stream/storage/level/local/writer/format/csv/fastcsv/fastcsv.go

```diff
@@ -0,0 +1,54 @@
+// Package fastcsv provides parallel writer/formatter for CSV records.
+//   - There is WritersPool with fixed number of writers.
+//   - The io.Writer.Write method of the out writer is always called with the whole CSV row.
+//   - There is no built-in synchronization of the output, so the io.Writer.Write method must be protected externally.
+//   - The values are always quoted, so it is possible to generate row in single pass, without scanning the column value.
+//   - The resulting CSV is therefore a bit larger, but we rely on the compression that may follow in the writers chain.
+package fastcsv
+
+import (
+	"io"
+	"runtime"
+	"sync"
+)
+
+type WritersPool struct {
+	sem  chan struct{}
+	pool *sync.Pool
+}
+
+func NewWritersPool(out io.Writer, writers int) *WritersPool {
+	if writers <= 0 {
+		writers = runtime.GOMAXPROCS(0)
+	}
+
+	p := &WritersPool{}
+	p.pool = &sync.Pool{
+		New: func() any {
+			return newWriter(out)
+		},
+	}
+
+	p.sem = make(chan struct{}, writers)
+	for i := 0; i < writers; i++ {
+		p.sem <- struct{}{}
+	}
+
+	return p
+}
+
+func (p *WritersPool) WriteRow(cols *[]any) error {
+	// The algorithm below is more efficient than
+	// if we send a pointer to a free writer directly through the channel.
+	//
+	// The sync.Pool is internally implemented using per-processor local pools.
+	// When goroutine is scheduled to run on a specific thread associated with specific processor
+	// and will try to retrieve an object from the pool, sync.Pool will first look in the current processor local pool.
+	// Read more: https://victoriametrics.com/blog/tsdb-performance-techniques-sync-pool/
+	<-p.sem
+	w := p.pool.Get().(*writer)
+	err := w.WriteRow(cols)
+	p.pool.Put(w)
+	p.sem <- struct{}{}
+	return err
```

Comment:
`p.sem` is used to limit number of parallel writers (to number of CPUs, by default).

`struct{}` is a token, number of the tokens in the channel buffer match number of free writers

Reply:
```
Using sync.Pool to reuse objects in I/O-bound tasks is much less efficient than re-using objects for CPU-bound tasks.
```
We actually use it in writer. So I would expect that this has more `I/O` operations than CPU ones.

Reply:
I don't understand, there is no I/O in `fastcsv` Go package at all. 

It needs CPU and memory buffer per row (which is re-used) - so it's almost only about the CPU.

Reply:
Let's have discussion about this.

Reply:
@Matovidlo please add comments, rows order, etc.